Add ability to search for method/field usage and use custom update sites

jvz opened the pull request:
This adds three new features:

* A new option to override the update center URL to fetch plugins from.
* A new option to search for qualified method names used by plugins (minus the rest of the method signature).
* A new option to search for qualified field names used by plugins.

Example output from the example given in USAGE.adoc (and a field lookup):

<img width="1022" alt="image" src="https://user-images.githubusercontent.com/791275/84702606-898ebe80-af1c-11ea-8464-059b53672fd7.png">

@jeffret-b @Wadeck @daniel-beck

daniel-beck commented on src/main/java/org/jenkinsci/deprecatedusage/Options.java:
```diff
@@ -16,36 +24,53 @@
+    public File additionalMethodsFile;
+
+    @Option(name = "--onlyAdditionalMethods", depends = "-M", usage = "Only include in the report the specified methods")
+    public boolean onlyAdditionalMethods;
```
It seems to me that the `onlyAdditional*` options could be collapsed into one. Or is there a real use case so specify some, but neither all nor none of them?

This would be a minor breaking change, but not introduce a bunch of addition options that realistically are only ever coming in a group.

jvz replied:
Changed to `--onlyIncludeSpecified` with the short option `-i`.

jvz commented:
Major update applied. Note that the work directory layout has changed slightly in order to support multiple update centers.

jeffret-b commented on USAGE.adoc:
```diff
@@ -50,3 +50,22 @@ Writing output/deprecated-and-unused.html
+=== Finding Usage of Methods and Fields
+
+The command line flag `-M filename.txt` or `--additionalMethods filename.txt` allows specific methods to be searched for.
+The additional flag `-i` or `--onlyIncludeSpecified` will filter the output report to only show usage of the methods specified.
```
nit:
```suggestion
Use the additional flag `-i` or `--onlyIncludeSpecified` to only show usage of the methods specified.
```

jeffret-b commented on USAGE.adoc:
```diff
@@ -50,3 +50,22 @@ Writing output/deprecated-and-unused.html
+
+The command line flag `-M filename.txt` or `--additionalMethods filename.txt` allows specific methods to be searched for.
+The additional flag `-i` or `--onlyIncludeSpecified` will filter the output report to only show usage of the methods specified.
+For example, to find usage of `hudson.util.Secret.toString`, the following line should be included in the specified file:
```
nit:
```suggestion
For example, to find usage of `hudson.util.Secret.toString`, include the following line in the specified file:
```

jeffret-b commented on USAGE.adoc:
```diff
@@ -50,3 +50,22 @@ Writing output/deprecated-and-unused.html
+hudson/util/Secret#toString
+----
+
+Class names should be specified in their fully qualified form (i.e., inner classes should be referenced via `Outer$Inner` rather than by their canonical names).
```
More nit:
```suggestion
Specify class names in their fully qualified form (i.e., inner classes should be referenced via `Outer$Inner` rather than by their canonical names).
```

jeffret-b commented on USAGE.adoc:
```diff
@@ -50,3 +50,22 @@ Writing output/deprecated-and-unused.html
+----
+
+Class names should be specified in their fully qualified form (i.e., inner classes should be referenced via `Outer$Inner` rather than by their canonical names).
+Method names should be specified after the hash.
```
```suggestion
Specify method names after the hash.
```

jeffret-b commented on USAGE.adoc:
```diff
@@ -50,3 +50,22 @@ Writing output/deprecated-and-unused.html
+
+Class names should be specified in their fully qualified form (i.e., inner classes should be referenced via `Outer$Inner` rather than by their canonical names).
+Method names should be specified after the hash.
+Note that you cannot specify which method overload signature, but they are included in reports.
```
Not sure I'm getting this one right.
```suggestion
You cannot specify distinct method signatures. All variants are included in reports.
```

jeffret-b commented on USAGE.adoc:
```diff
@@ -50,3 +50,22 @@ Writing output/deprecated-and-unused.html
+Method names should be specified after the hash.
+Note that you cannot specify which method overload signature, but they are included in reports.
+
+The command line flag `-F filename.txt` or `--additionalFields filename.txt` allows specific field usages to be found.
```
```suggestion
The command line flag `-F filename.txt` or `--additionalFields filename.txt` allows you to find specific field usages.
```

jeffret-b commented on USAGE.adoc:
```diff
@@ -50,3 +50,22 @@ Writing output/deprecated-and-unused.html
+
+The command line flag `-F filename.txt` or `--additionalFields filename.txt` allows specific field usages to be found.
+The syntax here is the same as for specifying method names.
+The additional flag `-i` will filter the output report to only show specified fields.
```
Actually, this needs to be pulled out and discussed as it applies to all operations generally and not repeated for each different type.

jeffret-b commented:
Forgot to add. I tested this with a variety of inputs. I didn't get very fancy with specifying methods and fields. I tried several different basic searches. The more complicated ones should work just fine based on standard specifications. I tried it with a different update center and that worked fine.

daniel-beck commented:
There seems to be no deduplication of core releases. Given the following command indicates the same weekly release of Jenkins at the same URL is analyzed twice:

    mvn exec:java -Dexec.args='--updateCenters=https://updates.jenkins.io/2.176/update-center.json,https://updates.jenkins.io/update-center.json'

daniel-beck commented on src/main/java/org/jenkinsci/deprecatedusage/Main.java:
```diff
@@ -1,75 +1,167 @@
+                futures.add(plugin.downloadIfNotExists(client, threadPool).handle((success, failure) -> {
+                    concurrentDownloadsPermit.release();
+                    if (failure != null) {
+                        System.out.println("Error downloading " + plugin);
```
Shouldn't this rather be a URL?

    Error downloading work/59a92455/ibm-application-security/1.3.2/ibm-application-security.hpi

Doesn't seem all that useful.

jvz replied:
Ha, yeah, the logging in this could all be a little improved it seems. Thanks for the ideas!

daniel-beck commented:
I'm at about 1.5% of downloads failing. This indicates that we should have a retry feature, or at the very list a summary at the end of output listing which plugins failed to download, with a note that re-running the tool will re-attempt downloads.

daniel-beck commented on src/main/java/org/jenkinsci/deprecatedusage/JenkinsFile.java:
```diff
@@ -65,80 +52,92 @@ public String getWiki() {
+                    if (checksum == null || checksum.matches(data)) {
+                        Files.write(file, data, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
+                    } else {
+                        future.completeExceptionally(new DigestException("Invalid checksum for downloaded file " + file));
```
```
Error downloading work/6ba2a29d/libvirt-slave/1.8.6/libvirt-slave.hpi
java.security.DigestException: Invalid checksum for downloaded file work/6ba2a29d/libvirt-slave/1.8.6/libvirt-slave.hpi
```

```
$ cat work/6ba2a29d/libvirt-slave/1.8.6/libvirt-slave.hpi
cat: work/6ba2a29d/libvirt-slave/1.8.6/libvirt-slave.hpi: No such file or directory
```

Output is at least misleading here.

jvz replied:
Damn relative directories! Yeah, fairly misleading since the script doesn't even tell you that it supports an optional CLI arg for specifying the output directory (defaults to creating a temp directory printed at the beginning). Will improve log messages.

daniel-beck replied:
@jvz The problem here is that the update center advertises a file not yet on mirrors. It's a long standing infra issue. Still shouldn't complain about a checksum of a file that doesn't exist.

jvz replied:
I've improved the log messages a bit. Let me know if you still see anything weird.

daniel-beck replied:
This still happens. When a file cannot be download (somewhat common due to INFRA-160), it still complains about an invalid checksum (which should never happen and is a serious problem).

jvz replied:
I added some manual code to retry a download when an invalid checksum is detected. I think this might be caused by a truncated download that doesn't cause a connection reset or other exception. The updated code will retry up to three times when invalid digests are found. This is a separate retry strategy from the HttpClient one.

jvz commented:
> There seems to be no deduplication of core releases. Given the following command indicates the same weekly release of Jenkins at the same URL is analyzed twice:
> 
> ```
> mvn exec:java -Dexec.args='--updateCenters=https://updates.jenkins.io/2.176/update-center.json,https://updates.jenkins.io/update-center.json'
> ```
> 
> ```
> Analyzing deprecated APIs in work/59a92455/core/2.243/jenkins.war
> Analyzing deprecated APIs in work/6ba2a29d/core/2.243/jenkins.war
> Finished deprecated API analysis in work/6ba2a29d/core/2.243/jenkins.war
> Finished deprecated API analysis in work/59a92455/core/2.243/jenkins.war
> ```
> 
> There's also no deduplication of plugin downloads. As the majority of plugins will be identical between common update centers, that's a pretty big waste of time, and the support for multiple update sites brings minimal benefit over two separate executions.
> 
> I think the assumption that, between update sites, identical versions of the same component are the same is reasonable, and could be used to speed this up.

I can certainly return and try to implement this again, but I had previously given up on the approach as writing multithreaded code to mass download potentially overlapping things was getting fairly non-trivial. My "complicated" approach that seemed to work was to acquire FileChannel locks on the output files so that two download requests for the same plugin don't clobber one another. Alternatively, I can figure out some sort of processing queue to deduplicate the HTTP requests in the first place so that file locking won't be necessary. One of these approaches should work.

daniel-beck commented:
@jvz I would look at the JSONs as a whole to determine what to download, and if any number of them tell you to download `$plugin-$version` you only do that once and reuse. Existing `$plugin/$version/` directories would be good enough, maybe remove the code that deletes obsolete directories there.

jvz commented:
Yeah, I think that could work. I'll have some changes ready later for further review.

jvz commented:
I'll note that some of the errors you'll see during downloads are because either the plugin download file's checksum doesn't actually match the published checksum, or some artifacts seem to be corrupted. The HttpClient library I added performs automatic retries by default, hence why this PR seems to generally work better now at downloading everything in one go. I could potentially change this to simply emit warnings for plugins that don't match the update center metadata (any of which I'd expect to fix itself with a new plugin release).

daniel-beck commented:
> I'll note that some of the errors you'll see during downloads are because either the plugin download file's checksum doesn't actually match the published checksum, or some artifacts seem to be corrupted

Most of them (all except 1) were "Connection reset by peer" and gone by the third execution of the same command.

Just one, the error I quoted in a line comment, was a very recent release not yet on mirrors.

jvz commented:
> > I'll note that some of the errors you'll see during downloads are because either the plugin download file's checksum doesn't actually match the published checksum, or some artifacts seem to be corrupted
> 
> Most of them (all except 1) were "Connection reset by peer" (or similar, don't exactly remember) and gone by the third execution of the same command.
> 
> Just one, the error I quoted in a line comment, was a very recent release not yet on mirrors.

Alright, I'll take a closer look while refactoring.

daniel-beck commented:
Still failing 20/1800 plugin downloads, more than 1% without an explanation 😦 

If the degree of concurrency is too high, maybe allow manually overriding that? IIUC, the concurrency is based on CPUs, but uses bandwidth as resource, which seems weird.

Not a huge problem, just not great.

```
Error downloading http://updates.jenkins-ci.org/download/plugins/azure-dev-spaces/3.0.0/azure-dev-spaces.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/azure-dev-spaces/3.0.0/azure-dev-spaces.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
…
Error downloading http://updates.jenkins-ci.org/download/plugins/azure-keyvault/2.0/azure-keyvault.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/azure-keyvault/2.0/azure-keyvault.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
…
Error downloading http://updates.jenkins-ci.org/download/plugins/changelog-history/1.7/changelog-history.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/changelog-history/1.7/changelog-history.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
…
Error downloading http://updates.jenkins-ci.org/download/plugins/configuration-as-code-groovy/1.1/configuration-as-code-groovy.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/configuration-as-code-groovy/1.1/configuration-as-code-groovy.hpi
Error downloading http://updates.jenkins-ci.org/download/plugins/config-rotator/2.0.1/config-rotator.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/config-rotator/2.0.1/config-rotator.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
Error downloading http://updates.jenkins-ci.org/download/plugins/cucumber-living-documentation/2.2.2/cucumber-living-documentation.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/cucumber-living-documentation/2.2.2/cucumber-living-documentation.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
…
Error downloading http://updates.jenkins-ci.org/download/plugins/generic-webhook-trigger/1.67/generic-webhook-trigger.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/generic-webhook-trigger/1.67/generic-webhook-trigger.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
…
Error downloading http://updates.jenkins-ci.org/download/plugins/gitlab-oauth/1.6/gitlab-oauth.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/gitlab-oauth/1.6/gitlab-oauth.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
…
Error downloading http://updates.jenkins-ci.org/download/plugins/ezwall/0.3/ezwall.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/ezwall/0.3/ezwall.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
…
Error downloading http://updates.jenkins-ci.org/download/plugins/heavy-job/1.1/heavy-job.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/heavy-job/1.1/heavy-job.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
…
Error downloading http://updates.jenkins-ci.org/download/plugins/hyper-commons/0.1.5/hyper-commons.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/hyper-commons/0.1.5/hyper-commons.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
…
Error downloading http://updates.jenkins-ci.org/download/plugins/google-git-notes-publisher/0.3/google-git-notes-publisher.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/google-git-notes-publisher/0.3/google-git-notes-publisher.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
…
Error downloading http://updates.jenkins-ci.org/download/plugins/kubernetes-client-api/4.9.2-2/kubernetes-client-api.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/kubernetes-client-api/4.9.2-2/kubernetes-client-api.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
…
Error downloading http://updates.jenkins-ci.org/download/plugins/jdk-tool/1.4/jdk-tool.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/jdk-tool/1.4/jdk-tool.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
…
Error downloading http://updates.jenkins-ci.org/download/plugins/jenkins-design-language/1.23.2/jenkins-design-language.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/jenkins-design-language/1.23.2/jenkins-design-language.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
…
Error downloading http://updates.jenkins-ci.org/download/plugins/label-linked-jobs/5.1.2/label-linked-jobs.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/label-linked-jobs/5.1.2/label-linked-jobs.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
…
Error downloading http://updates.jenkins-ci.org/download/plugins/no-agent-job-purge/1.2/no-agent-job-purge.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/no-agent-job-purge/1.2/no-agent-job-purge.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
…
Error downloading http://updates.jenkins-ci.org/download/plugins/pipeline-multibranch-defaults/2.1/pipeline-multibranch-defaults.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/pipeline-multibranch-defaults/2.1/pipeline-multibranch-defaults.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
…
Error downloading http://updates.jenkins-ci.org/download/plugins/pry/1.1/pry.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/pry/1.1/pry.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
…
Error downloading http://updates.jenkins-ci.org/download/plugins/testweaver/1.0.6/testweaver.hpi -> /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/testweaver/1.0.6/testweaver.hpi
org.apache.hc.core5.http.ConnectionClosedException: Connection closed by peer
```

daniel-beck commented:
Behavior with multiple update sites tested with
```
$ cat classes.txt 
org/apache/http/client/methods/HttpPost
```
and
```
mvn exec:java -Dexec.args='--updateCenters=https://updates.jenkins.io/2.176/update-center.json -C classes.txt -i'
mvn exec:java -Dexec.args='--updateCenters=https://updates.jenkins.io/2.176/update-center.json,https://updates.jenkins.io/update-center.json -C classes.txt -i'
```

Usage in `audit-trail` is new in a release requiring Jenkins newer than 2.176.x, so the first call doesn't list it, the second does.

jvz commented:
Concurrency is controllable via `-D number` or `--downloadConcurrent number`

daniel-beck commented:
It would be nice if the log output contained explicitly the values chosen for some arguments so I can be sure it's effective. `-D 5` doesn't feel much different from the default of 48 😆

jvz commented:
Ah, I think I figured out the issue for the parallelism. I was using a ForkJoinPool that used the available processors as its default limit. The way I've written this code would make more sense to simply use a cached thread pool instead of a work stealing one as this is mostly waiting on IO in an asynchronous API that isn't based on CompletableFuture in the first place. This seems to greatly increase performance correlating with the `-D` value. I'll push some updates around that.

jvz commented:
Ooh, and I can add some manual code to handle retrying the HTTP request for `ConnectionClosedException` as that doesn't seem to be included in the library's auto-retry strategy.

jvz commented:
Alright, few small changes since your last review, @daniel-beck:

* Changed from a work-stealing thread pool (with parallelism=Ncpus) to a cached thread pool which makes more sense with a `Semaphore` since they're blocking on threads. (The equivalent way of doing that in a `ForkJoinPool` involves some other APIs I didn't feel like figuring out right now which likely won't improve the performance of this all that much)
* Only prints a warning message when the checksum of a downloaded file doesn't match.
* HTTP retry strategy configuration updated to use up to 3 retries spaced out by five seconds; this seemed to be sufficient in my local testing to get the super fat plugins to download more reliably.
* Changed the log message for connection timeouts to specify that it gave up trying to download that URL; other errors will still show the generic exception error message.

daniel-beck commented:
WARNING! Checksum of http://updates.jenkins-ci.org/download/plugins/echarts-api/4.8.0-2/echarts-api.hpi does not match update center value!

Huh. Well, let's see.

```
$ curl -LI http://updates.jenkins-ci.org/download/plugins/echarts-api/4.8.0-2/echarts-api.hpi
HTTP/1.1 302 Found
Date: Thu, 09 Jul 2020 21:54:50 GMT
Server: Apache/2.4.29 (Ubuntu)
Location: http://mirrors.jenkins-ci.org/plugins/echarts-api/4.8.0-2/echarts-api.hpi
Content-Type: text/html; charset=iso-8859-1

HTTP/1.1 302 Found
Date: Thu, 09 Jul 2020 21:54:50 GMT
Server: Apache/2.4.29 (Ubuntu)
X-MirrorBrain-Mirror: archives.jenkins-ci.org
X-MirrorBrain-Realm: other
Link: <http://mirrors.jenkins-ci.org/plugins/echarts-api/4.8.0-2/echarts-api.hpi.meta4>; rel=describedby; type="application/metalink4+xml"
Link: <http://archives.jenkins-ci.org/plugins/echarts-api/4.8.0-2/echarts-api.hpi>; rel=duplicate; pri=1; geo=us
Location: http://archives.jenkins-ci.org/plugins/echarts-api/4.8.0-2/echarts-api.hpi
Content-Type: text/html; charset=iso-8859-1

HTTP/1.1 404 Not Found
Date: Thu, 09 Jul 2020 21:54:50 GMT
Server: Apache/2.4.29 (Ubuntu)
Content-Type: text/html; charset=iso-8859-1
```

Still an inappropriate message, now even more panic inducing 😦

daniel-beck commented:
I am almost certain that the retry strategy isn't used. Even with ridiculously large numbers of retries and intervals, the failed downloads appear in the log immediately (as the plugin names are iterated in alphabetical order), which makes no sense.

In fact, I created an anonymous subclass of `DefaultHttpRequestRetryStrategy` to log some stuff, and there isn't a single call to `#retryRequest` with a `execCount` that isn't `1`.

jvz commented:
You did remember to compile it, right? `mvn exec:java` doesn't recompile the code when you change things; I learned that the hard way recently :)

daniel-beck commented:
Yes, and my changes were visible, otherwise I wouldn't have claimed

> In fact, I created an anonymous subclass of `DefaultHttpRequestRetryStrategy` to log some stuff, and there isn't a single call to `#retryRequest` with a `execCount` that isn't `1`.


```
25195  09.07.20 23:50:15 /Applications/apache-maven-3.6.2/bin/mvn clean verify
25196  09.07.20 23:50:32 /Applications/apache-maven-3.6.2/bin/mvn exec:java -Dexec.args='--updateCenters=https://updates.jenkins.io/2.176/update-center.json -C classes.txt -i'
25197  09.07.20 23:50:57 rm -rf work/
25198  09.07.20 23:51:01 /Applications/apache-maven-3.6.2/bin/mvn exec:java -Dexec.args='--updateCenters=https://updates.jenkins.io/2.176/update-center.json,https://updates.jenkins.io/update-center.json -C classes.txt -i'
…
25200  10.07.20 00:03:40 /Applications/apache-maven-3.6.2/bin/mvn clean verify
25201  10.07.20 00:03:49 /Applications/apache-maven-3.6.2/bin/mvn exec:java -Dexec.args='--updateCenters=https://updates.jenkins.io/2.176/update-center.json,https://updates.jenkins.io/update-center.json -C classes.txt -i'
25202  10.07.20 00:09:36 rm -rf work/
25203  10.07.20 00:09:40 /Applications/apache-maven-3.6.2/bin/mvn clean verify
25204  10.07.20 00:09:47 /Applications/apache-maven-3.6.2/bin/mvn exec:java -Dexec.args='--updateCenters=https://updates.jenkins.io/2.176/update-center.json,https://updates.jenkins.io/update-center.json -C classes.txt -i'
25205  10.07.20 00:13:08 /Applications/apache-maven-3.6.2/bin/mvn clean verify
25206  10.07.20 00:13:13 /Applications/apache-maven-3.6.2/bin/mvn exec:java -Dexec.args='--updateCenters=https://updates.jenkins.io/2.176/update-center.json,https://updates.jenkins.io/update-center.json -C classes.txt -i'
25207  10.07.20 00:13:52 /Applications/apache-maven-3.6.2/bin/mvn clean verify
25208  10.07.20 00:13:57 /Applications/apache-maven-3.6.2/bin/mvn exec:java -Dexec.args='--updateCenters=https://updates.jenkins.io/2.176/update-center.json,https://updates.jenkins.io/update-center.json -C classes.txt -i'
…
25210  10.07.20 00:21:04 rm -rf work/
25211  10.07.20 00:21:08 /Applications/apache-maven-3.6.2/bin/mvn clean verify
25212  10.07.20 00:21:19 /Applications/apache-maven-3.6.2/bin/mvn exec:java -Dexec.args='--updateCenters=https://updates.jenkins.io/2.176/update-center.json,https://updates.jenkins.io/update-center.json -C classes.txt -i'
25213  10.07.20 00:21:59 /Applications/apache-maven-3.6.2/bin/mvn clean verify
25214  10.07.20 00:22:04 /Applications/apache-maven-3.6.2/bin/mvn exec:java -Dexec.args='--updateCenters=https://updates.jenkins.io/2.176/update-center.json,https://updates.jenkins.io/update-center.json -C classes.txt -i'
25215  10.07.20 00:24:39 /Applications/apache-maven-3.6.2/bin/mvn clean verify
25216  10.07.20 00:24:44 /Applications/apache-maven-3.6.2/bin/mvn exec:java -Dexec.args='--updateCenters=https://updates.jenkins.io/2.176/update-center.json,https://updates.jenkins.io/update-center.json -C classes.txt -i'

jvz commented:
I think I figured out the problem with the retry strategy. It seems I misinterpreted the default exception list there as the ones to retry on, but those are the ones _not_ to retry on! Seems like we'd actually like to retry on at least one of these scenarios.

jvz commented:
Alright, who can merge this now?
